util: protect against monkeypatched Object prototype for inspect()

Trott opened the pull request:
Prevent affects of monkeypatching (for example) Object.keys() when calling util.inspect().

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/20604/

Trott commented:
> We create an artificial error that can only happen due to Object.keys being monkey patchable.
> 
> If we prevent that (which should IMHO be done at some point)

Do you mean prevent `Object.keys()` monkey-patching, or do you mean prevent this internal module from using a monkey-patchable `Object.keys()` (by, say, caching/saving the original value of `Object.keys()` before a user is able to get in and change it)?

Assuming the latter, I think this is a good test to have until we actually do that, and then it can be removed. What this test does in its current form is test that the guard-against-spec-changes (and also guard-against-V8-bugs) is working as expected.

Trott commented:
@BridgeAR Ooh, actually perhaps a thing we could both be happy with is changing the guard code to an assertion. So once we've cached `Object.keys()`so that it isn't monkey-patchable, the code might look like this (with necessary line-wrapping added):

```js
try {
      keys = Object.keys(value);
    } catch (err) {
      assert(isNativeError(err) &&  err.name === 'ReferenceError' && isModuleNamespaceObject(value));
      keys = Object.getOwnPropertyNames(value);
    }      
```

BridgeAR commented:
@Trott I am fine switching to the assertion as you suggested. We could even log the actual error (just passing through the error as message).

That could also be done without preventing Object.keys from being monkey patchable here. That way the line will be properly covered without actually changing the behavior.

Trott commented:
@BridgeAR OK, I've updated this to avoid using a monkey-patched `Object` in `inspect.js` and modified the test to confirm that nothing breaks when `Object` is monkey-patched. And I've converted the guard code to use the new internal micro-`assert()`. PTAL!

@addaleax This is a significant enough change from the one you approved that you may wish to re-review?

Trott commented:
Defensively adding `semver-major` in case this somehow breaks APMs or something. (Seems unlikely, but weirder stuff has happened. Happy to remove `semver-major` if that's the consensus.) 

/ping @nodejs/tsc (for reviews, because: `semver-major`)

addaleax commented on lib/internal/util/inspect.js:
```diff
@@ -227,7 +229,7 @@ inspect.colors = Object.assign(Object.create(null), {
 
 // Don't use 'blue' not visible on cmd.exe
-inspect.styles = Object.assign(Object.create(null), {
+inspect.styles = primordials.Object.assign(Object.create(null), {
```
I think you also want to change the `.create()` call?

(Is there a chance this might be better done through destructuring `primordials`?)

Trott replied:
Argh, how did I miss that?

Trott replied:
(Looks like I missed a few....)

Trott replied:
Instead of destructuring individual functions, I went with `const { Object } = primordials;'. That keeps the diff down, prevents other `Object` stuff from creeping in, and allows an easy way to add `Map` or whatever else we want going forward. On the other hand, not sure if people will object to locally masking `Object` that way. Seems like a *good* idea to me in this context, but let's see what y'all think....

addaleax replied:
Yeah no, that’s exactly what I had in mind :)

Trott commented:
@bmeck @ljharb

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/20675/

ljharb commented on lib/internal/util/inspect.js:
```diff
@@ -63,6 +63,11 @@ const {
+const assert = require('internal/assert');
+
+// Avoid monkey-patched built-ins.
+const { Object } = primordials;
```
this won't prevent `delete Object.keys`. Try:
```suggestion
const { Object: { keys: objectKeys, getOwnPropertyNames: gOPNs } } = primordials;
```

ljharb commented on lib/internal/util/inspect.js:
```diff
@@ -383,13 +388,9 @@ function getKeys(value, showHidden) {
     try {
       keys = Object.keys(value);
```
and then here:
```suggestion
      keys = objectKeys(value);
```

ljharb commented on lib/internal/util/inspect.js:
```diff
@@ -383,13 +388,9 @@ function getKeys(value, showHidden) {
-      }
+      assert(isNativeError(err) && err.name === 'ReferenceError' &&
+             isModuleNamespaceObject(value));
+      keys = Object.getOwnPropertyNames(value);
```
```suggestion
      keys = gOPNs(value);
```

addaleax commented:
@ljharb These are *not* the original builtins, they are frozen copies that are not accessible to userland.

ljharb commented:
@addaleax ahh that's the context i'm missing, thanks.

Trott commented:
Resume Build CI: https://ci.nodejs.org/job/node-test-pull-request/20678/

Trott commented:
Landed in 1847696

Trott commented:
Ugh, forgot a CITGM run. Sorry, everyone. Here's one after-the-fact (run against master since this landed):  https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1734/
